Abgabe Exercise #2

m-oehme opened the pull request:
We changed a lot of things that arent necessarily part of exercise 2.
The reason is that we had issues with setting up Vue-Test-Utils with the given folder structure.
Also it seems that having all the code in one index.js file was a bad idea in exercise 1. @6csharpmajor had issues setting up in Windows environment.

Because of all this restructuring und configuration issues we could only do a bare bones testing of our components.

Commit:
set up jest example tests

Commit:
Merge branch 'master' into development

Commit:
Merge branch 'master' into development

Commit:
Merge branch 'master' into development

Commit:
Merge branch 'master' into development

Commit:
added badge for CI status

Commit:
Merge branch 'master' into development

Commit:
changing to vue components files cuz I thing tseting works only with these

Commit:
Had lots of problems with having everything in one index file.
Moved everything into sepate src folders and using Component Files now!

Adding correct config for Jest/Vue testing. Official online tutorials are NOT correct!!

Commit:
renamed file

Commit:
changed add button

On Add todo shows edting field instead of "new todo"

Commit:
added testin files

Commit:
added build passing badge for dev branch

Commit:
fixed babel config for testing

Commit:
added some simple tests for vue

Commit:
added prettier

Commit:
added test for todolist component

Commit:
added more test

vctosp committed:
removed canvas

Penomatikus commented on tests/unit/App.spec.js:
```diff
@@ -0,0 +1,9 @@
+describe("App", () => {
+  test("is a Vue instance", () => {
+    const wrapper = mount(App);
+    expect(wrapper.isVueInstance()).toBeTruthy();
```
Nice, didn't know, that was possible!

Penomatikus commented:
Good work on that, for the first vue app/test! Didn't know you could test for a vue instance it self! Well done!

m-oehme commented:
@Penomatikus thx for the review. The instance check we found somewhere in the guidelines for vue-test-utils

@roschaefer In the exercise descriptions you have pictures of the log output for the Tests with jest and vue-test-utils. On your pics you have log of all the steps for each test. Ours only show the end result "Test Passed".
Is there some config where missing?

Commit:
changed jest to verbose mode to get more detail on tests

m-oehme commented:
@roschaefer never mind my previous question about showing more detail for tests.
Found that you can use jest cli options for vue-test-utils all the same.

roschaefer commented on src/components/TodoItem.vue:
```diff
@@ -0,0 +1,51 @@
+            delToDo: function() {
+                this.isEditing = false;
+                this.$emit("deleteEntry", this.index);
+                // this.$parent.deleteEntry(this.index);
```
```suggestion
```

I know it's nitpicking. But checking in commented code confuses other developers. I googled and found https://kentcdodds.com/blog/please-dont-commit-commented-out-code there the author calls it "cognitive load".

m-oehme replied:
Sry I didn't  notice it. 
Is it better to use `this.$parent.deleteEntry` or `this.$emit("deleteEntry"` in this particular case?

roschaefer replied:
I didn't know `this.$parent` searched the internet and found https://vuejs.org/v2/guide/components-edge-cases.html

I saw that the title is "handling edge cases" and that's enough for me to say: You should really `this.$emit` here, because the parent component can decide on its own what to do with the event.

`this.$parent.doSomething` hard-wires the child component with the parent component, because that way it can only live inside a parent component that has a method `deleteEntry`.

m-oehme replied:
ok thank you.
We changed it back to using `$emit` already.

roschaefer commented on src/components/TodoItem.vue:
```diff
@@ -0,0 +1,51 @@
+            save: function() {
+                this.isEditing = false;
+                this.todo.message = this.temp;
+                this.$emit("saveEntry", this.temp, this.index);
```
```suggestion
                this.$emit("saveEntry", this.todo , this.index);
```

It shouldn't be the List's job to put the object together.

m-oehme replied:
Using `this.todo` will throw an exception for me
`TypeError: Converting circular structure to JSON`
but we will change it to:
```suggestion
                this.$parent.saveEntry(this.todo, this.index);
```

m-oehme replied:
We changed it in the next commit.

roschaefer commented on src/components/TodoList.vue:
```diff
@@ -0,0 +1,50 @@
+        { id: 2, message: "Bar", isEditing: false },
+        { id: 3, message: "Baz", isEditing: false }
+      ],
+      lastID: 3
```
Avoid state! It's the root of all evil.

m-oehme replied:
we changed it now

roschaefer commented on src/components/TodoList.vue:
```diff
@@ -0,0 +1,50 @@
+    };
+  },
+  methods: {
+    addEntry: function() {
```
```suggestion
    addEntry() {
```

roschaefer commented on src/components/TodoList.vue:
```diff
@@ -0,0 +1,50 @@
+    addEntry: function() {
+      this.lastID++;
+      this.todoListData.push({
+        id: this.lastID,
```
If your object has the `id`, why does the `TodoItem` component emits both the index and the todo? Wouldn't just the todo be enough?

m-oehme replied:
Did you make the wrong code in your suggestion?

When saving we emit the index because it is much more efficient to search via index then having to look for the right id first.

We changed it to use id instead now.

roschaefer replied:
I think I made the suggestion before I learned that your `todo` object has a property `id`. In that case I would suggest to put all the things together

Am I right that you emit the todo along with a redundant id as another parameter?

m-oehme replied:
We changed it to emit todo object now (after the first code review)
Before we indeed each TodoItem received a Todo and an autogenerated Index.

when emiting a method we used the index and todo. so the Id was redundant.

But like I said we fixed that.

roschaefer commented on tests/unit/TodoItem.spec.js:
```diff
@@ -0,0 +1,45 @@
+    describe("edit text and submit", () => {
+      it("$emits save with edited todo", () => {
+        // wrapper.find("input").setValue("edited.todo.text");
+        wrapper.find("#save").trigger("click");
```
Uh-oh. You have non-atomic tests, don't you? Execution order matters? That's bad.

You should use a `beforeEach` and do
```js
wrapper.find("#edit").trigger("click");
```
there.

That way, the wrapper gets initialized (and clicked and what not) before each test case in the `describe` block. 

https://jestjs.io/docs/en/setup-teardown

m-oehme replied:
I don't know what you mean by "non-atomic test"?
We put the button trigger test as it was explained here: https://vue-test-utils.vuejs.org/guides/#common-tips

We also wanted to test if the data was actually changed but couldn't figure out how. And the guides from vue-test-utils didn't help

roschaefer replied:
Here is a nice explanation of atomic tests: https://defragdev.com/blog/?p=714

What I mean here: You are altering wrapper. This `wrapper` will survive until the next test case and have the save button clicked.

You can achieve atomic test be re-initializing the `wrapper ` in a `beforeEach`:
https://jestjs.io/docs/en/setup-teardown

roschaefer commented on tests/unit/TodoList.spec.js:
```diff
@@ -0,0 +1,34 @@
+
+describe("TodoList", () => {
+  it("contains 3 items", function() {
+    expect(wrapper.findAll(TodoItem).length).toBe(3);
```
```suggestion
    expect(wrapper.findAll(TodoItem)).toHaveLength(3);
```

https://jestjs.io/docs/en/expect#tohavelengthnumber

roschaefer commented on tests/unit/TodoList.spec.js:
```diff
@@ -0,0 +1,34 @@
+    expect(wrapper.emitted()).toBeTruthy();
+  });
+
+  // describe("delete item from todolist", () => {
```
Delete commented code. Or do you have troubles to set it up? Just ask and I'll help.

roschaefer reviewed (changes requested):
Well done!

![well done](https://i.giphy.com/media/lvOnlEYunAwOkHjgmU/giphy.webp)

But I see that your linter is missing. Please add one more command in `.travis.yml` so I can see that you know how to configure your build server. Another group has added test coverage which is also fine.

- [x] setup build server and status checks
- [x] check linter on build server

Commit:
Update tests/unit/TodoList.spec.js

Co-Authored-By: Robert Schäfer <git@roschaefer.de>

Commit:
Update src/components/TodoList.vue

Co-Authored-By: Robert Schäfer <git@roschaefer.de>

Commit:
fixed linter for travis ci

forgot to add correct configuration for eslint

used most config of the config from (but edited) https://github.com/SyJuly/Systems-Development-and-Frameworks/blob/frontend_testing/1/ToDoApp/.eslintrc.json

Commit:
Merge remote-tracking branch 'origin/development' into development

Commit:
problem with old snapshot file

snapshot file let tests fail

Commit:
changed todoitem method calls to parent

on save a todoitem will now emit itself instead of its changed value

Commit:
removed commented code

Commit:
added editmode property

statemanagement is cool now

Commit:
removed index

list now uses its ids instead of indexes

Commit:
added default message

Commit:
changed back to using $emit

$parent.method... seams to be type safe but it throws error in test cases

Commit:
Merge branch 'master' into development

roschaefer commented:
> @roschaefer never mind my previous question about showing more detail for tests.
> Found that you can use jest cli options for vue-test-utils all the same.

Yes, `jest` is the test runner, `vue-test-utils` is just a library that you can use from within your jest tests to conveniently mount components in a non-browser environment. Find all of `jest`'s cli options here: https://jestjs.io/docs/en/cli
